### PR TITLE
[FLINK-12192] [table-planner-blink] Add support for generating optimized logical plan for grouping sets and distinct aggregate

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Add support for generating optimized logical plan for grouping sets and distinct aggregate*

## Brief change log

  - *extends AggregateExpandDistinctAggregatesRule to fix some bugs*
  - *adds DecomposeGroupingSetsRule to rewrite aggregate with grouping sets into  a regular aggregate with expand*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added GroupingSetsTest that validates queries with grouping sets/cube/rollup*
  - *Added DistinctAggregateTest that validates queries with distinct aggregate*
  - *Addes DecomposeGroupingSetsRuleTest that validates DecomposeGroupingSetsRule*
  - *Addes FlinkAggregateExpandDistinctAggregatesRuleTest that validates FlinkAggregateExpandDistinctAggregatesRule*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by chunweilei** [comment]
> extends AggregateExpandDistinctAggregatesRule to fix some bugs
Would you please explaining a bit more about the bugs?

**Comment by godfreyhe** [comment]
> > extends AggregateExpandDistinctAggregatesRule to fix some bugs
> 
> Would you please explaining a bit more about the bugs?

Bugs and unsupported cases include
1. Throws an exception if an aggregate contains both approximate distinct aggregate call and accurate distinct aggregate call.
2. Excludes non-simple aggregate(e.g. CUBE, ROLLUP).
3. Fix bug: Some aggregate functions (e.g. COUNT) has a non-null result even without any input.
4. Fix bug: Add filter argument into rewritten aggregateCall if its filter argument is not -1.

which are also recorded in the class comments.

**Comment by chunweilei** [comment]
Got it. Appreciate your detailed explanation much.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelOptUtil.scala [hunk 1]
```diff
@@ -104,10 +105,19 @@ object FlinkRelOptUtil {
     *
     * @return Whether any of the aggregates are accurate DISTINCT
     */
-  def containsAccurateDistinctCall(aggCalls: Seq[AggregateCall]): Boolean = {
+  def containsAccurateDistinctCall(aggCalls: util.List[AggregateCall]): Boolean = {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelOptUtil.scala]: move to `AggregateUtil`?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelOptUtil.scala [hunk 2]
```diff
@@ -104,10 +105,19 @@ object FlinkRelOptUtil {
...
     aggCalls.exists(call => call.isDistinct && !call.isApproximate)
   }
 
+  /**
+    * Returns whether any of the aggregates are approximate DISTINCT.
+    *
+    * @return Whether any of the aggregates are approximate DISTINCT
+    */
+  def containsApproximateDistinctCall(aggCalls: util.List[AggregateCall]): Boolean = {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelOptUtil.scala]: ditto

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelOptUtil.scala (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelOptUtil.scala]: done

**Comment by KurtYoung** [comment]
There is an error:

**Comment by godfreyhe** [comment]
@KurtYoung the failed test is fixed

**Pull request closed without merging**